### PR TITLE
Sync ``libasr`` from LPython

### PR DESCRIPTION
**Pull request by czgdp1807**
Corresponding LPython PR - https://github.com/lcompilers/lpython/pull/2148

### CODE REVIEW

File: src/libasr/codegen/asr_to_wasm.cpp [hunk 1]
```diff
@@ -720,7 +720,9 @@ class ASRToWASMVisitor : public ASR::BaseVisitor<ASRToWASMVisitor> {
             if (ASR::is_a<ASR::Function_t>(*item.second)) {
                 ASR::Function_t *s =
                     ASR::down_cast<ASR::Function_t>(item.second);
-                this->visit_Function(*s);
+                if (ASRUtils::get_FunctionType(s)->n_type_params == 0) {
+                    this->visit_Function(*s);
+                }
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_wasm.cpp:725]: To be reverted (https://github.com/lfortran/lfortran/commit/a8b6b9a96c962fa1f2e620c6fbfb5375750e9c37#diff-9309bfe433b6c942f7e9c397b85275ec0acb7d076c40775bdf61cf2272be951a).

File: src/libasr/codegen/asr_to_wasm.cpp [hunk 2]
```diff
@@ -754,7 +756,7 @@ class ASRToWASMVisitor : public ASR::BaseVisitor<ASRToWASMVisitor> {
                 nullptr, 0, nullptr, 0, x.m_body, x.n_body, nullptr,
                 ASR::abiType::Source, ASR::accessType::Public,
                 ASR::deftypeType::Implementation, nullptr, false, false, false, false, false,
-                false, false, false);
+                nullptr, 0, nullptr, 0, false, false, false);
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_wasm.cpp:759]: To be reverted (TBR) (https://github.com/lfortran/lfortran/commit/a8b6b9a96c962fa1f2e620c6fbfb5375750e9c37#diff-9309bfe433b6c942f7e9c397b85275ec0acb7d076c40775bdf61cf2272be951a)

File: src/libasr/codegen/asr_to_wasm.cpp [hunk 3]
```diff
@@ -2896,13 +2898,6 @@ class ASRToWASMVisitor : public ASR::BaseVisitor<ASRToWASMVisitor> {
         handle_print(x);
     }
 
-    void visit_StringFormat(const ASR::StringFormat_t &x) {
-        diag.codegen_warning_label(
-            "StringFormat not implemented yet, ignored for now",
-            {x.m_fmt->base.loc}, "ignored");
-        this->visit_expr(*x.m_fmt);
-    }
-
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_wasm.cpp:2905]: TBR (To be reverted) (https://github.com/lfortran/lfortran/commit/6261b7042072f73585c8eab4bc0ffa044523638b)

File: src/libasr/codegen/asr_to_c_cpp.h [hunk 4]
```diff
@@ -1726,33 +2173,6 @@ R"(#include <stdio.h>
...
-                out += src;
-                if (i < x.n_args-1) {
-                    out += ", ";
-                }
-            }
-        }
-        out += "}";
-        src = out;
-    }
-
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_c_cpp.h:1755]: TBR (https://github.com/lfortran/lfortran/commit/8af388ad5fb8e15406045ad800e6a1718adf6aee)

File: src/libasr/codegen/asr_to_c_cpp.h [hunk 5]
```diff
@@ -2322,15 +2851,6 @@ R"(#include <stdio.h>
...
     }
-
-    void visit_IntrinsicFunctionSqrt(const ASR::IntrinsicFunctionSqrt_t &x) {
-        std::string out = "sqrt";
-        headers.insert("math.h");
-        this->visit_expr(*x.m_arg);
-        out += "(" + src + ")";
-        src = out;
-    }
-
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_c_cpp.h:2333]: TBR (https://github.com/lfortran/lfortran/commit/8af388ad5fb8e15406045ad800e6a1718adf6aee)

File: src/libasr/pass/loop_vectorise.cpp [hunk 6]
```diff
@@ -129,13 +129,6 @@ class LoopVectoriseVisitor : public PassUtils::SkipOptimizationFunctionVisitor<L
         ASR::expr_t* loop_start = x.m_head.m_start;
         ASR::expr_t* loop_end = x.m_head.m_end;
         ASR::expr_t* loop_inc = x.m_head.m_increment;
-        LCOMPILERS_ASSERT(loop_start);
-        LCOMPILERS_ASSERT(loop_end);
-        if (!loop_inc) {
-            int a_kind = ASRUtils::extract_kind_from_ttype_t(ASRUtils::expr_type(x.m_head.m_v));
-            ASR::ttype_t *type = ASRUtils::TYPE(ASR::make_Integer_t(al, x.base.base.loc, a_kind));
-            loop_inc = ASRUtils::EXPR(ASR::make_IntegerConstant_t(al, x.base.base.loc, 1, type));
-        }
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/loop_vectorise.cpp:138]: TBR (https://github.com/lfortran/lfortran/commit/5703fad12f44453e58f571ea433cc3e5156371a9).

File: src/libasr/pass/print_arr.cpp [hunk 7]
```diff
@@ -103,10 +103,15 @@ class PrintArrVisitor : public PassUtils::PassVisitor<PrintArrVisitor>
...
-            if (PassUtils::is_array(x.m_values[i])) {
+            // TODO: This will disallow printing array pointer in Fortran
+            // Pointers are treated the same as normal variables in Fortran
+            // However, LPython prints the address of pointers when you do
+            // print(some_pointer). Same goes for C/C++ (if we add their frontends in future).
+            // So we need to figure out a way to de-couple printing support from libasr
+            // or add nodes according to the frontends because each frontend will have a different
+            // way of handling printing of pointers and non-pointers
+            if (!ASR::is_a<ASR::Pointer_t>(*ASRUtils::expr_type(x.m_values[i])) &&
+                PassUtils::is_array(x.m_values[i])) {
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/print_arr.cpp:114]: TBR (https://github.com/lfortran/lfortran/commit/4f5ac0f0379efb6c69e8066de3a3f3c99d86ab85).

File: src/libasr/pass/print_arr.cpp [hunk 8]
```diff
@@ -203,10 +208,15 @@ class PrintArrVisitor : public PassUtils::PassVisitor<PrintArrVisitor>
...
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/print_arr.cpp:219]: TBR (https://github.com/lfortran/lfortran/commit/4f5ac0f0379efb6c69e8066de3a3f3c99d86ab85#diff-fec4a77919fb1c47061183081472565540714093b4f88cd2c4e56716b1ec3cc0).

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 9]
```diff
@@ -1686,15 +1372,15 @@ LFORTRAN_API void _lfortran_read_int32(int32_t *p, int32_t unit_num)
     if (unit_num == -1) {
         // Read from stdin
         FILE *fp = fdopen(0, "r+");
-        (void)fread(p, sizeof(int32_t), 1, fp);
+        fread(p, sizeof(int32_t), 1, fp);
```
Comment (Shaikh-Ubaid) [inline src/libasr/runtime/lfortran_intrinsics.c:1375]: TBR (https://github.com/lfortran/lfortran/commit/749563226e871de13d766acbce68a92b5cc54b85#diff-057d851a7170d57e545f3cbb89a1ef0e902dabd51532e5c25cb8d363b8c78725).

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 10]
```diff
@@ -1686,15 +1372,15 @@ LFORTRAN_API void _lfortran_read_int32(int32_t *p, int32_t unit_num)
...
         fclose(fp);
         return;
     }
     if (!unit_to_file[unit_num]) {
         printf("No file found with given unit\n");
         exit(1);
     }
-    (void)fread(p, sizeof(int32_t), 1, unit_to_file[unit_num]);
+    fread(p, sizeof(int32_t), 1, unit_to_file[unit_num]);
```
Comment (Shaikh-Ubaid) [inline src/libasr/runtime/lfortran_intrinsics.c:1383]: TBR (https://github.com/lfortran/lfortran/commit/749563226e871de13d766acbce68a92b5cc54b85#diff-057d851a7170d57e545f3cbb89a1ef0e902dabd51532e5c25cb8d363b8c78725).

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 11]
```diff
@@ -1703,7 +1389,7 @@ LFORTRAN_API void _lfortran_read_char(char **p, int32_t unit_num)
         // Read from stdin
         *p = (char*)malloc(16);
         FILE *fp = fdopen(0, "r+");
-        (void)fread(*p, sizeof(char), 16, fp);
+        fread(*p, sizeof(char), 16, fp);
```
Comment (Shaikh-Ubaid) [inline src/libasr/runtime/lfortran_intrinsics.c:1392]: TBR (https://github.com/lfortran/lfortran/commit/749563226e871de13d766acbce68a92b5cc54b85#diff-057d851a7170d57e545f3cbb89a1ef0e902dabd51532e5c25cb8d363b8c78725).

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 12]
```diff
@@ -1712,7 +1398,7 @@ LFORTRAN_API void _lfortran_read_char(char **p, int32_t unit_num)
         exit(1);
     }
     *p = (char*)malloc(16);
-    (void)fread(*p, sizeof(char), 16, unit_to_file[unit_num]);
+    fread(*p, sizeof(char), 16, unit_to_file[unit_num]);
```
Comment (Shaikh-Ubaid) [inline src/libasr/runtime/lfortran_intrinsics.c:1401]: TBR (https://github.com/lfortran/lfortran/commit/749563226e871de13d766acbce68a92b5cc54b85#diff-057d851a7170d57e545f3cbb89a1ef0e902dabd51532e5c25cb8d363b8c78725).

**Comment by czgdp1807** [comment]
@certik LLVM 10 tests fail because in LPython we use `builder->CreateCmp` API which I think wasn't present in LLVM 10. So would it be possible to stop supporting LLVM 10 in LFortran (we don't support it in LPython anyways)?

**Comment by certik** [comment]
If it is easy to emulate, I would maybe keep LLVM 10, since we already have it setup. But if it is hard, then we can drop LLVM 10. 

As a rule of thumb, we need to support all versions of our main dependencies for the last 5 years. I would say the main dependencies are just LLVM, C++ and operating systems (re2c and Bison are minor, and we pregenerate them anyway). So we need to support Ubuntu from 5 years ago, and C++ compilers from 5 years ago. LLVM 10 was released on 24 Mar 2020. 5 years ago was LLVM 7, but it's really hard to support LLVM 7, since it didn't yet have the new JIT, so I would say that is not worth it. Another way to look at it is that we should be able to just build using Ubuntu 18.04 LTS ideally, but in practice it's quite hard, but at least Ubuntu 20.04 we should support. It looks like it still uses LLVM 10 as the default.

All in all, I would support LLVM 10 for now if it is at all possible. What does it involve to support it?

**Comment by czgdp1807** [comment]
Well LLVM 10 isn't supported in LPython. Even if I some how make LLVM 10 work with LFortran, things like this will keep happening in the future, because we haven't made sure in LPython that we use APIs which work with LLVM 10 as well.

**Comment by certik** [comment]
We have to test LLVM 11-16 in both LPython and LFortran anyway. I would test LLVM 10 - 16, I think it's worth it. Let's discuss the technical difficulties.

**Pull request merged**